### PR TITLE
Add scatter for CUDA support

### PR DESCRIPTION
**Pull request by yuehhua**
Transfer from FluxML/NNlib.jl#296.

### CODE REVIEW

File: src/scatter.jl [hunk 1]
```diff
@@ -0,0 +1,44 @@
+ATM_OPS = Dict((+) => CUDA.atomic_add!, (-) => CUDA.atomic_sub!, (max) => CUDA.atomic_max!, (min) => CUDA.atomic_min!,
+               (*) => CUDA.atomic_mul!, (/) => CUDA.atomic_div!, (&) => CUDA.atomic_and!, (|) => CUDA.atomic_or!)
+
+for (op, atm_op) in ATM_OPS
+    @eval function scatter_kernel!(op::typeof($(op)), dst, src, idx)
+        index = threadIdx().x + (blockIdx().x - 1) * blockDim().x
+
+        @inbounds if index <= length(idx)
+            i = Base._to_linear_index(dst, idx[index]...)
+            $(atm_op)(pointer(dst, i), src[index])
```
Comment (maleadt) [inline src/scatter.jl:10]: if you use `@atomic` you don't need to call `pointer` or calculate a linear index yourself

File: src/NNlibCUDA.jl [hunk 2]
```diff
@@ -4,6 +4,9 @@ using NNlib
 using CUDA
 using Random, Statistics
 
+const IntOrIntTuple = Union{Integer, NTuple{N,<:Integer} where N}
+const MAX_THREADS = 1024
```
Comment (maleadt) [inline src/NNlibCUDA.jl:8]: this isn't safe, querying the occupancy API would be better here

File: src/scatter.jl (hunk 1 above)
Comment (yuehhua) [reply src/scatter.jl]: Done.

File: src/NNlibCUDA.jl (hunk 2 above)
Comment (yuehhua) [reply src/NNlibCUDA.jl]: I modified it. Is it good now?

File: src/scatter.jl [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        else
+            dims_size = size(dst)[1:N]
+            max_dims_idx = prod(dims_size)
+            max_idx = max_dims_idx * length(idx)
+            op, dst, src, idx, dims, max_idx, max_dims_idx, dims_size
+        end
+
+        kernel = @cuda launch=false scatter_kernel!(args...)
+        config = launch_configuration(kernel.fun; max_threads=256)
+        threads = Base.min(max_idx, config.threads)
```
Comment (maleadt) [inline src/scatter.jl:35]: You don't need the `Base` prefix on 1.6 anymore.

File: src/scatter.jl [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+        @inbounds if index <= max_idx
+            j, k = divrem(index-1, max_dims_idx)
+            dims_i = CartesianIndices(dims_size)[k+1]
+            @atomic dst[Tuple(dims_i)..., idx[j+1]...] = $(op)(dst[Tuple(dims_i)..., idx[j+1]...], src[index])
+        end
+        return nothing
+    end
+
+    @eval function NNlib.scatter!(op::typeof($(op)), dst::CuArray{Tdst}, src::CuArray{Tsrc}, idx::CuArray{<:IntOrIntTuple}, dims::Val{N}) where {Tdst,Tsrc,N}
```
Comment (maleadt) [inline src/scatter.jl:22]: If you want to support wrapped arrays, it's best to use `AnyCuArray` here.

Comment (maleadt) [reply src/scatter.jl:22]: Also, do you need the `eval` here? Can't you just define this for all functions (and have `@cuda` throw a method error if using an unsupported op)?

File: src/scatter.jl (hunk 3 above)
Comment (yuehhua) [reply src/scatter.jl]: Done.

File: src/scatter.jl (hunk 4 above)
Comment (yuehhua) [reply src/scatter.jl]: Done.

**Comment by yuehhua** [comment]
Wait for merge of JuliaGPU/CUDA.jl#842

**Comment by yuehhua** [comment]
Now this PR relies on NNlib#master (or next patch release of v0.7.19) and CUDA#master (or next patch release of v3.0.4).

**Comment by maleadt** [comment]
Look for `cp.async` in `$CUDA/targets/x86_64-linux/include/cuda/std/barrier`.

**Comment by yuehhua** [comment]
@maleadt I have checked it. Is there somewhere to use asynchronous data movement? I didn't get your point.

File: src/scatter.jl [hunk 5]
```diff
@@ -0,0 +1,40 @@
+for op in [+, -, *, /, max, min, &, |]
+    @eval function scatter_kernel!(op::typeof($(op)), dst, src, idx)
```
Comment (CarloLucibello) [inline src/scatter.jl:2]:
```suggestion
    function scatter_kernel!(op, dst, src, idx)
```
why don't we define the kernel for generic ops and let it error out by itself for CUDA unsupported ones?

**Comment by maleadt** [comment]
Sorry, that was a comment on the wrong issue.

File: src/scatter.jl (hunk 5 above)
Comment (yuehhua) [reply src/scatter.jl]: Done.

**Comment by yuehhua** [comment]
@maleadt  Could you trigger the CI test again? I think it is ready to go.
All tests passed on my PC.

**Comment by maleadt** [comment]
You can always amend and force-push to re-trigger CI.

**Comment by CarloLucibello** [comment]
You can also close and reopen to retrigger

**Comment by yuehhua** [comment]
I need at least NNlib v0.7.19 and CUDA v3.0.4 to pass CI, but buildkite just have the following version.

```julia
[052768ef] CUDA v3.0.0
[f6369f11] ForwardDiff v0.10.18
[872c559c] NNlib v0.7.18
```

As I know, there should be compatible with current version.

**Comment by maleadt** [comment]
Then you need to bump the versions in the Manifest. Buildkite doesn't 'have' specific package versions.

**Comment by yuehhua** [comment]
Thank you. It works.

**Comment by CarloLucibello** [comment]
The compat bounds should also be updated

**Comment by yuehhua** [comment]
The compat bounds look somewhat like this?

**Comment by CarloLucibello** [comment]
And this other little piece is done, thank you for all these efforts!

**Comment by maleadt** [comment]
This shouldn't have been **merged**, but should have been squashed. Now there's 13 commits added to master for a single feature, many of which fail CI (complicating operations like `git bisect`).

**Pull request merged**